controlplane/kubeadm/controllers/kubeadm_control_plane_controller_test.go

```diff
@@ -0,0 +1,439 @@
+			objs: []runtime.Object{
+				kcp,
+				existingCACertSecret,
+			},
+			clusterName:  types.NamespacedName{Namespace: "test", Name: "foo"},
+			endpoint:     clusterv1.APIEndpoint{Host: "test.local", Port: 8443},
+			kcp:          kcp,
+			wantErr:      false,
+			createSecret: true,
+			expectSecret: true,
+		},
+		{
+			name: "Does not replace an existing kubeconfig if it already exist",
+			objs: []runtime.Object{
+				kcp,
+				existingKubeconfigSecret,
+			},
+			clusterName:  types.NamespacedName{Namespace: "test", Name: "foo"},
+			endpoint:     clusterv1.APIEndpoint{Host: "test.local", Port: 8443},
+			kcp:          kcp,
+			wantErr:      false,
+			createSecret: false,
+			expectSecret: true,
+		},
+		{
+			name: "Error if missing ca certificate secret",
+			objs: []runtime.Object{
+				kcp,
+			},
+			clusterName:  types.NamespacedName{Namespace: "test", Name: "foo"},
+			endpoint:     clusterv1.APIEndpoint{Host: "test.local", Port: 8443},
+			kcp:          kcp,
+			wantErr:      true,
+			createSecret: false,
+			expectSecret: false,
+		},
+		{
+			name: "Empty API endpoint returns early",
+			objs: []runtime.Object{
+				kcp,
+			},
+			clusterName:  types.NamespacedName{Namespace: "test", Name: "foo"},
+			endpoint:     clusterv1.APIEndpoint{},
+			kcp:          kcp,
+			wantErr:      false,
+			createSecret: false,
+			expectSecret: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fakeClient := fake.NewFakeClientWithScheme(setupScheme(), tt.objs...)
+			r := &KubeadmControlPlaneReconciler{
+				Client: fakeClient,
+				Log:    log.Log,
+			}
+			err := r.reconcileKubeconfig(context.Background(), tt.clusterName, tt.endpoint, tt.kcp)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("reconcileKubeconfig() error = %v, wantErr %v", err, tt.wantErr)
+			}
+
+			if tt.expectSecret {
+				kubeconfigSecret := &corev1.Secret{}
+				secretName := types.NamespacedName{
+					Namespace: "test",
+					Name:      secret.Name(tt.clusterName.Name, secret.Kubeconfig),
+				}
+				if err := r.Client.Get(context.Background(), secretName, kubeconfigSecret); err != nil {
+					t.Errorf("expected kubeconfig secret to exist: %v", err)
+				}
+				if tt.createSecret {
+					g.Expect(kubeconfigSecret.OwnerReferences).NotTo(gomega.BeEmpty())
+					g.Expect(kubeconfigSecret.OwnerReferences).To(gomega.ContainElement(*metav1.NewControllerRef(kcp, controlplanev1.GroupVersion.WithKind("KubeadmControlPlane"))))
+					g.Expect(kubeconfigSecret.Labels).To(gomega.HaveKeyWithValue(clusterv1.ClusterLabelName, tt.clusterName.Name))
+				} else {
+					g.Expect(kubeconfigSecret.Labels).To(gomega.Equal(existingKubeconfigSecret.Labels))
+					g.Expect(kubeconfigSecret.Data).To(gomega.Equal(existingKubeconfigSecret.Data))
+					g.Expect(kubeconfigSecret.OwnerReferences).NotTo(gomega.ContainElement(*metav1.NewControllerRef(kcp, controlplanev1.GroupVersion.WithKind("KubeadmControlPlane"))))
```

Comment:
do these have sufficient output indicating what fields are problematic without annotations:
`g.Expect(...., "some annotation")?`

we had such a general problem in k/k/test, but in this case it might be not needed.

Reply:
I haven't had much issue with them in the past, mostly because they point directly to the line that caused the error, we can always add annotations later if it starts to become an issue, though.